### ISSUE TITLE
Lua: structured concurrency, Promises, task pipelines

### ISSUE THREAD

**Issue by justinmk** [neovim/neovim#19624]
related: https://github.com/neovim/neovim/issues/11312

## Problem

There's no builtin abstraction for:

- Representing a task.
    - Example: `vim.system()` returns its own ad-hoc "task" that can be "awaited" via `:wait()`.
- Orchestrating "pipelines" (quasi monads?) of work ("tasks") and handling errors.
    - Example: shell-like task chains: `vim.fs.files():filter(..):map(function(f) vim.fs.delete(f) end)`
    - See also `vim.iter`.

## Expected behavior

- "Task" abstraction:
    - Maximally leveraging Lua coroutines + libuv. Only add concepts ("task", "promise") if absolutely needed.
    - Coroutines (or tasks that wrap coroutines) can be nested. ([ref](https://gregorias.github.io/posts/using-coroutines-in-neovim-lua/#addendum-whats-wrong-with-plenary-async))
    - Util to create an awaitable task from "normal" functions (cf. "promisify"?).
        - Don't want to call `await` everywhere, that is a horrible consequence of the JS async/await model. Instead consider Go's `go` for opt-in concurrency.
        - Can e.g. `vim.system()` be "promisified" without its knowledge? Or could it handle differently when it detects that it's in a coroutine? So `vim.system()` would be synchronous normally, but `vim.async(vim.system, ...):wait()` would be asynchronous.
            - See `cb_to_co()` from [this article](https://gregorias.github.io/posts/using-coroutines-in-neovim-lua/#callbackcoroutine-conversion).
    - Document (or generalize) ["coroutine to callback"](https://gregorias.github.io/posts/using-coroutines-in-neovim-lua/#coroutine-to-callback-conversion).
- Structured concurrency:
    - `await_all`, `await_any` (pseudo-names). See JS [Promise.all()](https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Global_Objects/Promise/all).
    - Tasks can be _canceled_.
    - Results (and failures) can be _aggregated_. (Can't do this with `jobwait()`!)
    - Failures/errors can be handled (possibly canceling the rest of the task tree).

## Implementations

- https://github.com/lewis6991/async.nvim
- https://github.com/nvim-neotest/nvim-nio "enables async I/O through higher level abstractions and utilities around coroutines"
- https://github.com/kevinhwang91/promise-async
- https://github.com/ms-jpq/lua-async-await

## Related

- https://github.com/sourcegraph/conc
- https://github.com/bitfield/script (example of "pipeline" interface)
- https://gregorias.github.io/posts/using-coroutines-in-neovim-lua/

### notes from [reddit discussion](https://old.reddit.com/r/neovim/comments/1g8lutp/coroutine_tutorial_for_neovim_lua/lt0rp9m/)

> Briefly speaking, without cancellability, a structured concurrency API can and should be based on fire-and-forget coroutine functions (fafcf). Anything else would likely either be reinventing the wheel or running into a non-composable mess that is currently Plenary’s async.
> 
> With pure fafcfs, you can have most of your requirements: ability to chain, start concurrent computations, wait for completion of arbitrary “fafcfs.” It’d be quite a expressive and elegant system.
> 
> As you notice, what’s missing is cancellability. That would require creating a special protocol that fafcfs need to conform to.
>
> After briefly looking at [async.nvim](https://github.com/lewis6991/async.nvim), it looks like it’s trying to build concurrency by stepping through chunks, which is kind of what Plenary’s async is trying to do.
> 
> Ideally, I’d like to try to achieve cancellability without resorting to reimplementing an event loop and a scheduler in Lua. Perhaps fafcfs with a special protocol for indicating that it shouldn’t proceed would be enough.
> 
> I believe that Neovim/Lua can take design hints from Python. In this regard, Lua fafcfs are [Python coroutines](https://docs.python.org/3/library/asyncio-task.html#awaitables). They can be nested freely. If you want to a cancellable computation, you need to use [Tasks](https://docs.python.org/3/library/asyncio-task.html#creating-tasks), which builds upon coroutines to provide a richer interface.

**Comment by clason** [comment]
I believe the gold standard is https://github.com/nvim-lua/plenary.nvim#plenaryasync, but @lewis6991 rolled his own for GitSigns. Either could conceivably be upstreamed if deemed useful enough.

**Comment by lewis6991** [comment]
For context here's the complete implementation Gitsigns uses:

```lua
local co = coroutine

local async_thread = {
   threads = {},
}

local function threadtostring(x)
   if jit then
      return string.format('%p', x)
   else
      return tostring(x):match('thread: (.*)')
   end
end

function async_thread.running()
   local thread = co.running()
   local id = threadtostring(thread)
   return async_thread.threads[id]
end

function async_thread.create(fn)
   local thread = co.create(fn)
   local id = threadtostring(thread)
   async_thread.threads[id] = true
   return thread
end

function async_thread.finished(x)
   if co.status(x) == 'dead' then
      local id = threadtostring(x)
      async_thread.threads[id] = nil
      return true
   end
   return false
end

local function execute(async_fn, ...)
   local thread = async_thread.create(async_fn)

   local function step(...)
      local ret = { co.resume(thread, ...) }
      local stat, err_or_fn, nargs = unpack(ret)

      if not stat then
         error(string.format("The coroutine failed with this message: %s\n%s",
         err_or_fn, debug.traceback(thread)))
      end

      if async_thread.finished(thread) then
         return
      end

      assert(type(err_or_fn) == "function", "type error :: expected func")

      local ret_fn = err_or_fn
      local args = { select(4, unpack(ret)) }
      args[nargs] = step
      ret_fn(unpack(args, 1, nargs))
   end

   step(...)
end

local M = {}

function M.wrap(func, argc)
   return function(...)
      if not async_thread.running() then
         return func(...)
      end
      return co.yield(func, argc, ...)
   end
end

function M.void(func)
   return function(...)
      if async_thread.running() then
         return func(...)
      end
      execute(func, ...)
   end
end

M.scheduler = M.wrap(vim.schedule, 1)

return M
```

**Comment by justinmk** [comment]
> I basically just copy and paste this for all my plugins now.

That's a good sign. Does it support cancellation? How does error handling look? How does it look to aggregate results?

**Comment by lewis6991** [comment]
> That's a good sign. Does it support cancellation? How does error handling look? How does it look to aggregate results?

It doesn't really support issuing out several coroutines with any kind of join, so there is nothing to cancel or aggregate, the 95% usecase is for use with sequential system commands (via `vim.loop.spawn`). We could implement that kind of stuff, I just haven't found a situation yet where I've really needed it.

> How does error handling look?

Here's an example of injecting `error("I'm an error")` mid-way through `setup()` in Gitsigns:
```
Error executing vim.schedule lua callback: ...e/pack/packer/start/gitsigns.nvim/lua/gitsigns/async.lua:67: The coroutine failed with this message: ...im/site/p
ack/packer/start/gitsigns.nvim/lua/gitsigns.lua:429: I'm an error
stack traceback:
        [C]: in function 'error'
        ...im/site/pack/packer/start/gitsigns.nvim/lua/gitsigns.lua:429: in function <...im/site/pack/packer/start/gitsigns.nvim/lua/gitsigns.lua:393>
stack traceback:
        [C]: in function 'error'
        ...e/pack/packer/start/gitsigns.nvim/lua/gitsigns/async.lua:67: in function <...e/pack/packer/start/gitsigns.nvim/lua/gitsigns/async.lua:62>
```

It just issues out both stacktraces: one from the application code, and one from the async lib. I've found it useful enough that I don't feel the need to improve it.

**Comment by williamboman** [comment]
I think this would be a great addition to the stdlib. I've been playing around with async implementations as well, and I'd like to bring forward some topics that I think are worth consideration:

- How to handle situations where the callback that resumes execution is being invoked more than once.
- Should callback signatures be standardized to allow easy translation between async <-> callback-style invocations? (e.g. `fun(..., callback: fun (err: any?, result: any?))`) For example:

  ```lua
  -- callback style
  vim.fs.find(".git", function (err, path)
    -- ...
  end)
  
  -- async style
  local a = vim.async
  local async_find = a.syncify(vim.fs.find) -- this could perhaps even be done automatically by vim.fs.find if it sees that no callback is passed as last arg
  local ok, path_or_err = pcall(async_find, ".git")
  ```
- How to handle nested coroutines? Should the async coroutine forward yields to parent coroutines, on what criterion?
- Should you be able to retrieve the result of an async function from a non-async context, e.g. through a callback?
- How to handle multiple return values across async/sync boundaries (`return val1, val2`)?
- Perhaps too early to even think about but I wanted to surface the idea anyway: Should user scripts perhaps be executed in a "global" async coroutine context? This would allow direct access to async primitives without users having to explicitly create async constructs.

**Comment by ii14** [comment]
> How to handle situations where the callback that resumes execution is being invoked more than once.

A closure could be created that can be called only once.

> `fun(..., callback: fun (err: any?, result: any?))`
> How to handle multiple return values across async/sync boundaries (`return val1, val2`)?

Maybe instead it could be `ok: boolean, results: any...`.

> How to handle nested coroutines? Should the async coroutine forward yields to parent coroutines, on what criterion?

Maybe I'm missing something, but it should always resume the caller?

**Comment by mfussenegger** [comment]
> What are the current patterns for queuing promises in Lua, that are being used in the wild for Nvim plugins (and Nvim core itself)? Are these good enough? Do we need any sugar in the Nvim lua stdlib to encourage these patterns?

In some of my plugins I started using the pattern norcalli mentioned in https://github.com/neovim/neovim/issues/11312#issuecomment-548042444:

```
local function send_request_sync(method, params)
	local co = coroutine.running()
	send_request(method, params, function(...) coroutine.resume(co, ...) end)
	return coroutine.yield()
end
```

and found that it's both simple and effective in avoiding callback nesting.

I think if we were to add some async/await or promise abstraction it would help to formulate the problem first, to help ensure we're solving the right thing.
E.g. I can see how functionality to race two or more operations, cancellation or waiting for multiple operations can be useful and the LSP stuff in neovim could probably benefit for some of it. But if all it does is add some sugar over the coroutine pattern above then I'd say it's not worth it.

**Comment by glepnir** [comment]
simple implementation of aysnc await is easy and can works well.
```lua
function async(f, ...)
    local co = coroutine.create(f)
    local function exec(...)
        local ok, data = coroutine.resume(co, ...)
        if not ok then
            error(debug.traceback(co, data))
        end
        if coroutine.status(co) ~= "dead" then
            data(exec)
        end
    end
    exec(...)
end

function await(f)
    local co = coroutine.running()
    local ret
    f(function(...)
        if coroutine.status(co) == "running" then
            ret = table.pack(...)
        else
            return coroutine.resume(co, ...)
        end
    end)
    if ret then
        return table.unpack(ret, 1, ret.n)
    else
        return coroutine.yield()
    end
end
```
if want to support async/await in builtin .  I think it should be more complete . support range , promise and status will be better

```lua
-- make ranged for loop 
range(start, stop, [step])
--promise
promis.new(function(resolve,reject))

--chain call
promise:chain_call(func)

-- cache
promis:catch(function)

-- await
promise:await()

-- resolve
promise:resolve(...)

-- error
promise:error(err)

-- status: Pending | Resolved | Errored
promise:status()
promise:isPending, promis:isResolved,promis:IsErroed

all/any waitForAll/waitForAny
```

**Comment by muniter** [comment]
Perhaps ideas and implementation from https://github.com/kevinhwang91/promise-async can be useful.

**Comment by Chromosore** [comment]
I've read something similar about Javascript: [A Study on Solving Callbacks with JavaScript Generators](https://archive.jlongster.com/A-Study-on-Solving-Callbacks-with-JavaScript-Generators), hope it might help.

**Comment by hrsh7th** [comment]
I am also implementing async-await for my plugin development. (I plan to use it primarily for sequencing code related to key mapping.)

Currently, I have not written the Promise.all equivalent, but will be able to add it easily.

https://github.com/hrsh7th/nvim-kit/blob/main/lua/___plugin_name___/Async/init.spec.lua#L27

**Comment by justinmk** [comment]
> I think if we were to add some async/await or promise abstraction it would help to formulate the problem first, to help ensure we're solving the right thing.
> E.g. I can see how functionality to race two or more operations, cancellation or waiting for multiple operations can be useful and the LSP stuff in neovim could probably benefit for some of it. But if all it does is add some sugar over the coroutine pattern above then I'd say it's not worth it.

@mfussenegger Absolutely. If there are patterns we can document without adding sugar, that is definitely preferred. So to close this, we can document those patterns. Or both: maybe we only need sugar for "race two or more operations, cancellation or waiting for multiple operations".

**Comment by wbthomason** [comment]
Just chiming in: `packer` has another async/await implementation based on https://github.com/ms-jpq/lua-async-await, but I consider it to be messy and would love to rip it out in favor of a stdlib alternative. It has limited support for interrupting a sequence of jobs and aggregating results, but could be improved in its API, robustness to errors, etc.

**Comment by oberblastmeister** [comment]
I have a lot of experience with this because I implemented the async library in plenary. The first iteration was basically a direct copy of https://github.com/ms-jpq/lua-async-await. It was not the best design because there is no point for async await with thunks. It felt like the extra thunks were just there to mirror the syntax from `javascript` or `rust`. `await` was literally implemented like this:
```lua
function await(f)
  f(nil)
end
```
It was also much harder to pass async functions to higher order functions.

The new version did not need `async` or `await`, and you can call functions just like normal synchronous functions. Furthermore, higher order functions that expect synchronous functions just work like normal. There is also no extra syntax for for loops
```lua
-- now iterators just work like normal
for entry in readdir() do
end

-- instead of
local iter = await(readdir())
while true do
  local entry = await(iter())
  if entry == nil then break end
end
```

Back then I didn't realize that this was basically delimited continuations. Now, I think we should just implement algebraic effects using one-shot delimited continuations which is described in this [paper]( http://logic.cs.tsukuba.ac.jp/~sat/pdf/tfp2020-postsymposium.pdf) and implemented in [eff.lua](https://github.com/Nymphium/eff.lua). The issue with just using coroutines for converting callback based functions is that you don't get to use coroutines for anything else. With algebraic effects, you can have other effects in addition to async effects. Algebraic effects can implement  `Reader` (local configuration), `Exceptions`, `State` (local state), `Generators`, and more. Also, algebraic effects separate the effects from interpretation. For example, imagine some program using the file system effect, log effect, and read effect:

```lua
local handle = handler {
  val = function(v) return v end,
  [Log] = function(k, arg)
    print(arg)
    return k()
  end,
  [ReadFile] = function(k, path)
    return k(fs.read_file(path))
  end,
  [Read] = function(k)
    return k(config)
  end
}
handle(my_program)
```
If you wanted to test your functions, you can change the interpretation to not use IO.
```lua
local vfs = {my_file = "asdfasdf", another_file = "adfadsfdf"}
local logs = {}

-- handle without IO!
local handle = handler {
  val = function(v) return v end,
  [Log] = function(k, arg)
    table.insert(logs, arg)
    return k()
  end,
  [ReadFile] = function(k, path)
    return k(vfs[path])
  end,
  [Read] = function(k)
    return k(config)
  end
}
```
Async could be implemented like this:
```lua
local handle = handler {
  val = callback,
  [Async] = function(k, fn, ...)
    fn(k, ...)
  end
}
```
Exceptions:
```lua
local handle = handler {
  val = function(v) return v end,
  [Exception] = function(k, e)
    return e
  end,
}
```

**Comment by hrsh7th** [comment]
I implemented `AsyncTask` (which is almost like the JavaScript Promises interface, but always synchronous if possible).

If neovim core supports async-await, I'm wondering if neovim should introduce promsie like async primitives or not.

- async-await
    - https://github.com/hrsh7th/nvim-kit/tree/main/lua/___kit___/kit/Async
- IO module (example of usage)
    - https://github.com/hrsh7th/nvim-kit/blob/main/lua/___kit___/kit/IO/init.lua#L114
- Convert callback style to Promise style
    - https://github.com/hrsh7th/nvim-kit/blob/main/lua/___kit___/kit/IO/init.lua#L54

**Comment by hrsh7th** [comment]
I found the interesting PR.
https://github.com/luvit/luv/pull/618

**Comment by lewis6991** [comment]
@oberblastmeister I've taken a look at eff.lua, and it appears to be a completely generalized version of the current async library we're using. If you take eff.lua and the async implementation you mentioned and refactor enough, you get to the same async lib I implemented in gitsigns.

 It's quite interesting, but I'm not sure if we need something as generalized as that. The main need is for an async lib to better utilize libuv and all of it's callback based functions.

> The issue with just using coroutines for converting callback based functions is that you don't get to use coroutines for anything else.

eff.lua uses coroutines in the exact same way the current async lib does, so not sure this is true.

> I implemented `AsyncTask` (which is almost like the JavaScript Promises interface, but always synchronous if possible). (The `always synchronous if possible` is needed to support feedkeys handling.)
> 
> If neovim core supports async-await, I'm wondering if neovim should introduce promsie like async primitives or not.
> 
>     * async-await
>       
>       * [hrsh7th/nvim-kit@`main`/lua/___kit___/kit/Async/init.lua](https://github.com/hrsh7th/nvim-kit/blob/main/lua/___kit___/kit/Async/init.lua?rgh-link-date=2022-12-02T04%3A13%3A30Z)
> 
>     * Keymap module
>       
>       * [hrsh7th/nvim-kit@`main`/lua/___kit___/kit/Vim/Keymap.spec.lua#L28](https://github.com/hrsh7th/nvim-kit/blob/main/lua/___kit___/kit/Vim/Keymap.spec.lua?rgh-link-date=2022-12-02T04%3A13%3A30Z#L28)
> 
>     * IO module
>       
>       * [hrsh7th/nvim-kit@`main`/lua/___kit___/kit/IO/init.lua#L114](https://github.com/hrsh7th/nvim-kit/blob/main/lua/___kit___/kit/IO/init.lua?rgh-link-date=2022-12-02T04%3A13%3A30Z#L114)
> 
>     * Convert callback style to Promise style
>       
>       * [hrsh7th/nvim-kit@`main`/lua/___kit___/kit/IO/init.lua#L54](https://github.com/hrsh7th/nvim-kit/blob/main/lua/___kit___/kit/IO/init.lua?rgh-link-date=2022-12-02T04%3A13%3A30Z#L54)

It looks like this implementation [creates a new coroutine](https://github.com/hrsh7th/nvim-kit/blob/main/lua/___kit___/kit/Async/init.lua#L23) for *every* async function. This is a big drawback compared to the plenary/gitsigns implementations that uses a single coroutine for each context. It also requires `await` which isn't needed in the other implementations.

I've created https://github.com/lewis6991/async.nvim which I'm using as a reference for all the plugins I use async in. The library is small enough that (for now), it's ok to just vendor directly in plugins without needing it as a depedency.

**Comment by hrsh7th** [comment]
Yes. My implementation creates a coroutine for each asynchronous operation.
I did not think this would affect performance. I will investigate.

I also wonder that "it is the drawback that requires the call to await".

Because I believe the difference is where and when to yield asynchronous primitives.

For example, look at the following example

```lua
-- Return just AsyncTask
local function timeout_async(ms)
  return AsyncTask.new(function(resolve))
    vim.defer_fn(resolve, ms)
  end)
end

-- Return AsyncTask if it is running in the main coroutine, otherwise yield.
local function timeout(ms)
  local task = timeout_async(ms)
  if vim.async.in_context() then
    return task:await()
  end
  return task
end

vim.async.run(function())
  local s = uv.now()
  timeout(200) -- does not need to `await` call.
   assert.is_true(math.abs(uv.now() - s - 200) < 10)
end)
```

In this example, `timeout_async` just returns an AsyncTask, but `timeout` delegates resolution to any async coroutine context.

I think it is a difficult question as to whether we should choose Promise, thunk, or another means as an asynchronous primitive.
However, it seems natural to me to introduce some kind of asynchronous primitive.

To me, the Promise style  seem like a good choice, since there are many JavaScript users and they are easy to understand.
Also, it is easy to convert parallel waits instead of serial ones. Promise.all(tasks):await()

Translated with www.DeepL.com/Translator (free version)

**Comment by justinmk** [comment]
> the Promise style seem like a good choice, since there are many JavaScript users and they are easy to understand.

Comforting javascript users is not a goal, and should not be factored into any of the designs here. As a typescript/javascript user myself, I'm very skeptical that async/await is a good interface. We should focus on what feels right for a Lua + Nvim.

> The main need is for an async lib to better utilize libuv and all of it's callback based functions.

Since the introduce of `jobstart()`, long before libuv was exposed. Network tasks like LSP are another common use-case.

**Comment by rcarriga** [comment]
> As a typescript/javascript user myself, I'm very skeptical that async/await is a good interface. We should focus on what feels right for a Lua + Nvim.

I've been using plenary's async module with neotest for a while now and have found the lack of an async/await syntax to be much more fitting with Lua's simplistic style of programming. IMO adding async/await would be clunky. 

> Sounds like https://github.com/lewis6991/async.nvim is informed by a lot of previous work, which is a good sign! Does it provide a way to cancel tasks, handle errors/cancellations, and aggregate results?

I wanted to introduce the async usage to nvim-dap-ui which is similar to LSP, where it's currently using a lot of callbacks. I worked off of @lewis6991's implementation and added the ability to cancel tasks and handle errors rather than them just being raised to the user. His implementation already covers aggregating results. It's very similar to Python's asyncio Task class just because I didn't want to have to design the API much. Also added some flow control primitives.

Implementation here https://github.com/rcarriga/nvim-dap-ui/blob/feat/new-api/lua/dapui/async/base.lua and some usage in tests https://github.com/rcarriga/nvim-dap-ui/blob/feat/new-api/tests/unit/async/base_spec.lua. Excuse the long namespaced names, it's just for generating documentation.

**Comment by rcarriga** [comment]
Thought I'd update as I've been having a bit of fun with this. I've done a fair bit of refactoring and adding of features in the last few days.

The main features are currently:
1. Handling results and errors of [tasks](https://github.com/rcarriga/nvim-dap-ui/blob/feat/new-api/lua/dapui/async/tasks.lua) along with cancelling.
2. Helper functions for coordinating tasks, including gathering results from a list of tasks and getting the result of the first task to complete (and cancelling the others).
3. Async versions of many libuv functions, pretty much the same way plenary does it with the added bonus of all of the functions being typed with emmylua annotations so users will have full type inference.
4. Creating a [fully documented and typed LSP client interface](https://github.com/rcarriga/nvim-dap-ui/blob/feat/new-api/lua/dapui/async/lsp-types.lua) which is generated by a script based off of the [LSP spec meta model](https://github.com/microsoft/vscode-languageserver-node/blob/main/protocol/metaModel.json)  and then implemented very simply with some metatables [here](https://github.com/rcarriga/nvim-dap-ui/blob/feat/new-api/lua/dapui/async/lsp.lua). 
5. Async primitives (queues, events and semaphores) which are often useful in async code, again based off of Python's asyncio modules.

Not sure if it's the direction core would want to go with an async library, (e.g. the LSP client but I was doing something similar for the DAP specification and LSP was mentioned as a key reason for the async usage so thought I'd slap something together for it). I think it does demonstrate the advantages of a single library  by implementing broadly useful features and providing a stable, tested solution for async that can be used by anybody.

Here is some sample usage of the LSP client. Open the script in the nvim-dap-ui branch to see the type checking/documentation with Lua LSP
```lua
local async = require("dapui").async

async
  .run(function()
    local client = async.lsp.client(2)

    -- Notifications usage
    client.notify.textDocument_willSave({
      reason = "3",
      textDocument = {
        uri = vim.uri_from_bufnr(0),
      },
    })

    -- Request usage
    local symbols = client.request.textDocument_documentSymbol(0, {
      textDocument = { uri = vim.uri_from_bufnr(0) },
    })

    return symbols
  end)
  .add_callback(function(task)
    if task.error() then
      vim.schedule_wrap(error)(task.error())
    else
      print(vim.inspect(task.result()))
    end
  end)
```

**Comment by lewis6991** [comment]
What is the purpose of `add_callback`. Why can't that code be appended to the end of the async function?

**Comment by rcarriga** [comment]
The `add_callback` usage is very arbitrary here. It's just to show result/error handling. I'd imagine the most common use case would be just for error handling. `add_callback` would be preferred over just wrapping everything in `pcall` because `pcall` won't catch errors outside of the coroutine such as cancellation.

**Comment by justinmk** [comment]
is `add_callback` more commonly called `then` ?

**Comment by rcarriga** [comment]
Ah yes that's true. Just named it based off of Python's [`add_done_callback`](https://docs.python.org/3.9/library/asyncio-task.html#asyncio.Task.add_done_callback), but I'd definitely prefer `then` :sweat_smile:

**Comment by justinmk** [comment]
> Edit: One key difference though is that this doesn't return a new task, so you can't chain them together like JS Promises

any reason we wouldn't want that?

**Comment by lewis6991** [comment]
I don't know about @justinmk thoughts, but I would like any implementation included in core to be fairly lean and minimal.

For the features I have focused on, this includes:
- [functions](https://github.com/lewis6991/async.nvim/blob/9bd4cbb69af3397b432c8274c36d30598b494c16/lua/async.lua#L61) for creating async contexts
- [functions](https://github.com/lewis6991/async.nvim/blob/9bd4cbb69af3397b432c8274c36d30598b494c16/lua/async.lua#L89) for creating async functions from callback style functions (libuv)
- [functions](https://github.com/lewis6991/async.nvim/blob/9bd4cbb69af3397b432c8274c36d30598b494c16/lua/async.lua#L104) for interleaving execution of multiple async functions (and aggregate the results).
- [Allow](https://github.com/lewis6991/async.nvim/blob/9bd4cbb69af3397b432c8274c36d30598b494c16/lua/async.lua#L89) async functions to run in a protected mode (like `pcall`) [WIP]
- Overall, async functions appearing like normal functions as much as possible.

And excludes:
- treating async functions like first class objects (tasks with methods)
- job cancellation: this can be done by providing custom callbacks to libuv functions instead of wrapping them directly
- chaining: just create a parent async function instead

Maybe we can revisit adding more advanced things in the future, but for now I think we should consider a minimum viable product that covers 95% of usecases with the least complexity.

**Comment by rcarriga** [comment]
> any reason we wouldn't want that?

Not particularly but I suppose the question would also be would we want it? Because we're only creating task objects for the top level async contexts instead of all async functions, I'm not sure there is a use case for chaining tasks. Happy to be wrong though! 

> I don't know about @justinmk thoughts, but I would like any implementation included in core to be fairly lean and minimal.

Totally makes sense, for core I think just getting a single implementation that can be built upon by plugins would be the minimal goal so that there is no issue mixing async plugins. After that it's a lot more subjective and could be done in a plugin. There is some work going on to get common dependencies like plenary added to rockspec so that users won't even have to specify the dependencies (For example https://github.com/nvim-neotest/neotest/pull/185) which would make keeping most of this outside core even more logical.

> - treating async functions like first class objects (tasks with methods)
> - job cancellation: this can be done by providing custom callbacks to libuv functions instead of wrapping them directly

I would question these exclusions. The task objects don't represent an async function but an entire async context (i.e. a running coroutine). Implementing error handling and cancelling outside of the core implementation reliably sounds like it would be difficult without manually wrapping async functions everywhere or users coupling their code to the core implementation (by using the coroutines directly) but I could definitely be missing something obvious.

The task objects are only some fairly minor changes over your base implementation. To me at least, they don't add much complexity in comparison to the payoff of being able to cancel, handle errors and get the traceback (useful for hung tasks).

**Comment by lewis6991** [comment]
> I would question these exclusions. The task objects don't represent an async function but an entire async context (i.e. a running coroutine). Implementing error handling and cancelling outside of the core implementation reliably sounds like it would be difficult without manually wrapping async functions everywhere or users coupling their code to the core implementation (by using the coroutines directly) but I could definitely be missing something obvious.

My opinion is completely based on the fact I've never needed these features. If I did, my instinct would be to construct things so I can get access to the underlying uv handle so I can explicitly call `close` to kill any jobs which may or may not call an exit callback. Or just not use the async lib at all. At the end of the day, the main motivation here is just to eliminate "callback-hell", but sometimes it isn't the worst solution depending on what you're trying to do.

> The task objects are only some fairly minor changes over your base implementation. To me at least, they don't add much complexity in comparison to the payoff of being able to cancel, handle errors and get the traceback (useful for hung tasks).

If that's the case, then we can definitely consider them.

**Comment by justinmk** [comment]
> Implementing error handling and cancelling outside of the core implementation reliably sounds like it would be difficult without manually wrapping async functions everywhere or users coupling their code to the core implementation

yeah, I see error handling and cancel as requirements. Doing that manually is unpleasant, it's one of the key features of a good concurrency interface.

**Comment by lewis6991** [comment]
Error handling is a slippery slope. We don't want to be implementing our own exception framework by setting a precedent here. A `pcall` like interface should be good enough.

Looking at the implementation, cancel looks totally reasonable. Not sure about the 8 other methods though, quite OOP.

**Comment by rcarriga** [comment]
> Looking at the implementation, cancel looks totally reasonable. Not sure about the 8 other methods though, quite OOP.

Yeah as I said the interface was basically copying python so OOP is right :sweat_smile:  It could be trimmed quite a bit if we wanted. For example something more `pcall`-like could be:
```lua
---@class Task
---@field cancel fun(): nil Cancels the task
---@field trace fun(): string|nil Get the current stack trace of the task (if running)

local task = async.run(function() end, function(success, err_or_results, trace)
  if not success then
    print(err_or_results .. "\n" .. trace)
    return
  end
  local result = unpack(err_or_results)
  -- Do something with result
end)
```
Adding the trace as an extra argument because once the coroutine dies, the `trace` function won't work.

We could even just return the cancel function directly but returning a table leaves it more open to extending in future. I think the trace function is useful. For example I'm using it in the tests to get useful tracebacks on timeouts which plenary's async library doesn't provide.

Also the above loses the ability to add multiple callbacks but not sure if it'd really be needed (I've not needed it so far)

**Comment by lewis6991** [comment]
Hmm, I would try and tweak this to something like:

```lua
local task = async.run(function() end)

-- Called in an async context
local success, result1_or_err, result2_or_trace = async.pcall(task)
-- You won't always care about the trace, or maybe not even the error so the typical use is
local success, result = async.pcall(task)  -- just like pcall

-- Called outside an async context
async.pcall(task, function(success, ...)
  if not success then
    local err, trace = ...
    print(err .. "\n" .. trace)
    return
  end
  local result = ...
  -- Do something with result
end)
```

That way a lot of these optional features are more opt-in and don't noise up the typical use case.

**Comment by oberblastmeister** [comment]
>> The issue with just using coroutines for converting callback based functions is that you don't get to use coroutines for anything else.

> eff.lua uses coroutines in the exact same way the current async lib does, so not sure this is true.

@lewis6991 I was just commenting about how you need some more logic if you want to use coroutines for anything other than async, for example generators.

**Comment by lewis6991** [comment]
I think @rcarriga implementation covers this by keeping track of what coroutines are used for async.

**Comment by oberblastmeister** [comment]
If you yield in a handler for generators, all your async stuff will be caught. Likewise, if you yield in a handler for async, all your generator stuff will be caught. I don't see how the implementation solves this. But, it depends on if we need support for other effects. If we only need async, then it is fine as is.

**Comment by lewis6991** [comment]
If the generator is contained completely within an async function, it should work just fine as the yield will go yield to the generators coroutine right? You can't however interleave them, this is just a limitation of using coroutines for async.

**Comment by oberblastmeister** [comment]
That is correct.

**Comment by hrsh7th** [comment]
Just sharing my thoughts for this discussion.

To be upfront, I'm not a stickler for the explicit async-await style of JavaScript.
But I think it's a safe and understandable way.

For example, gitsigns.nvim case.

[diff.tl](https://github.com/lewis6991/gitsigns.nvim/blob/main/teal/gitsigns/diff.tl#L4) looks like a normal function.
But in fact, it must be executed in an async-context, or else an error will be occur.
The reason is that `require('gitsigns.diff_int').run_diff` is an async-function and it's called in sync-style.

If we choose the `explicit await` style, the diff.tl will be the following form.

```lua
return function(a: {string}, b: {string}, linematch: boolean): {Hunk}
  local diff_opts = config.diff_opts
  local f: function({string}, {string}, string, boolean, integer): {Hunk}
  if diff_opts.internal then
    f = require('gitsigns.diff_int').run_diff
  else
    f = require('gitsigns.diff_ext').run_diff
  end

  local linematch0: integer
  if linematch ~= false then
    linematch0 = diff_opts.linematch
  end
  return f(a, b, diff_opts.algorithm, diff_opts.indent_heuristic, linematch0):await() -- HERE!
end
```

In this case, We are able to notice that `await is executed without async-context`.
This code still make error if the user that uses this function doesn't create async-context.

So I think it more readable to explicitly create the async-context at the point of the call when to use `await`.

```lua.
return function(a: {string}, b: {string}, linematch: boolean): {Hunk}
  return async.run(function() -- HERE!
    local diff_opts = config.diff_opts
    local f: function({string}, {string}, string, boolean, integer): {Hunk}
    if diff_opts.internal then
      f = require('gitsigns.diff_int').run_diff
    else
      f = require('gitsigns.diff_ext').run_diff
    end

    local linematch0: integer
    if linematch ~= false then
      linematch0 = diff_opts.linematch
    end
    return f(a, b, diff_opts.algorithm, diff_opts.indent_heuristic, linematch0):await() -- AND HERE!
  end)
end
````

If we adopt this approach, `async.run` must return some async-primitive (like Promise).
Otherwise, we will not be able to receive the result of executing this function in a sync-context.

Although, it's still the reasonable decision that the user must create async-context if they call async-function.

**Comment by lewis6991** [comment]
This is the same [concern](https://github.com/nvim-lua/plenary.nvim/pull/148#discussion_r638916337) I shared when the 2nd version of plenary's async lib was being worked on. The resolution was that we should just be more strict on naming conventions and/or annotate the code better with things like `-- @async`.

Admittedly this isn't something I've been very strict with in Gitsigns, but mainly because almost everything needs to run in an async context and runs jobs all over the place. In some places I've started adding `-- @async`.

In practice, it's not very different from things like `api-fast` or `textlock` which are in essence just other forms of context.

**Comment by hrsh7th** [comment]
That was it.

Similar to the need to properly handle textlock and fast-event contexts. It's insightful.

I agree that some sort of naming convention or having a dedicated place for async-functions might be a solution. (e.g.: vim.fs.async.open, vim.async.fs.open)

It simply became a matter of decision rather than concern for me.

**Comment by lewis6991** [comment]
> I'm not sure overloading a pcall function would make the most sense. 

I never suggested this. I suggested defining a specific `async.pcall` function that specificly handles the context for protected calls.

> If we just pass async functions directly, which would be common in async contexts then we have to decide how to handle arguments

This is exactly what I meant. Sorry for now being clear. I don't use anything like `async.run` and I wrap every single function, for me it's either `async.void` or `async.sync` (bad name I know).

> Which is great but if we want to allow calling it in a non async context then we'd either have to require the number of arguments to be provided or provide the callback before the arguments

`async.sync` and `async.wrap` require you to pass `argc`, so that's how we deal with that, and we wouldn't need to pass these to `pcall`. Passing argc to the wrappers is important and not something we can work around. This is a problem for Sumneko which has bad support for generics.

**Comment by rcarriga** [comment]
> I never suggested this. I suggested defining a specific async.pcall function that specificly handles the context for protected calls.

Sorry didn't think you meant overloading the built-in (that would be pretty nuts), was referring to overloading `async.pcall` depending on if it's an async context or not :sweat_smile:

> async.sync and async.wrap require you to pass argc, so that's how we deal with that, and we wouldn't need to pass these to pcall (async.run will need argc too). 

But you're not necessarily passing a wrapped function to `async.pcall`, it could just be a function that calls a wrapped function internally. 

```lua
local wrapped_func = async.wrap(function() return some_result end, 0)

local function async_func1()
    ...
    return wrapped_func()
end

local function async_func2()
    local success, some_result = async.pcall(async_func1)
end

```
I can't think of any way that would be able to reliably handle both supplying a callback (because you might want to supply callback in an async context) and supplying arguments without passing argc when calling it (it's essentially the same problem as `wrap` right?). 

I think I'm missing the use case of allowing `async.pcall` with a callback versus using `async.run`. Is it that you'd expect `async.run` to raise the error to the user if one occurred?

> Passing argc to the wrappers is important for details I won't go into and not something we can work around.

I understand why argc is passed to the wrapped functions, wasn't suggesting we'd change that.

> This is a problem for Sumneko which has bad support for generics.

Totally true just thought it was worth mentioning anyway.

**Comment by lewis6991** [comment]
> But you're not necessarily passing a wrapped function to async.pcall, it could just be a function that calls a wrapped function internally.

Ah right I see. Hmm, ok may need to think about this some more, but personally, I'm not too fussed about providing a callback to pcall. If it really comes to it then we may just have to pack args in a table, or, passing the callback in a curried style:

```lua
async.pcall(function(...) end)(...)(cb)
```

**Comment by lewis6991** [comment]
I don't think so.

**Comment by hrsh7th** [comment]
I think we need to consider `async-function` and `async-entrypoint` as two different things.

In the gitsigns.nvim example, `async-function` is `async.wrap(function() ... end)` and `async-entrypoint` is `async.void(function() ... end)`. 

There is no problem with the rule that `async-function(async.wrap)` can only be executed from an async-context. (It's just a design decision).

However, IMO, `async-entrypoint(async.void)` should provide the way to wait for the async completion.
(If so, we need to intoroduce `callback`, `thunk`, `Promise` or `Channel` for that)

If `async-entrypoint` does not provide some some of the way, the user will not be able to wait for async-function.

**Comment by lewis6991** [comment]
Gitsigns doesn't have a complete async implementation, just the things it needs. For a more complete implementation I'm using https://github.com/lewis6991/async.nvim as a staging area. It includes a couple more things:
- `async.sync`: (like `void`) but is intended to return its results via a callback. It returns a function of the signature `function(..., callback(results))`, so it works very much like a raw libuv function.
- `async.join`: to use with `async.sync` to run multiple functions concurrently.
- (Missing) job cancellation
- (Missing) `async.pcall`

> However, IMO, async-entrypoint(async.void) should provide the way to wait for the async completion.

How do you mean wait? You mean wait synchronously by blocking? That should be a pretty easy addition using `async.sync` with something like `uv.new_check`.

> If async-entrypoint does not provide some some of the way, the user will not be able to wait for async-function.

Do they need to be able to wait? Can't a user just always provide a callback?

**Comment by hrsh7th** [comment]
> How do you mean wait?

The `wait` means the `await`.
It can be achieve with `callback`, `thunk`, `Promise` etc.


Oh... I seems to mis-understanding the discussion...?

I think `async.pcall` isn't needed if the `async-entrypoint` provide some async-primitives.
(The `callback` can handle an error via `callback(err, res)` and also `Promise` can handle an error.)

The https://github.com/lewis6991/async.nvim seems to pass the arguments for the async-entrypoint.

> You mean wait synchronously by blocking?

No. I don't want to include `blocking-sync` into the neovim core. (At least initialy).

**Comment by lewis6991** [comment]
> I think async.pcall isn't needed if the async-entrypoint provide some async-primitives.
(The callback can handle an error via callback(err, res) and also Promise can handle an error.)

The idea is that `async.pcall` turns `callback(res)` into `callback(err, res)`.

**Comment by lewis6991** [comment]
I've updated https://github.com/lewis6991/async.nvim so:

- coroutines created by the async module are specially tracked so they can be lightly mixed with other coroutines.
- `async.void` and `async.create` (formerly `async.sync`) now return a handle with the methods `cancel(cb)` and `is_cancelled()` which is analogous to luv's `handle:close(cb)` and `handle:is_closing()`. It also gets access to a `co` field which is the coroutine in case anybody wants that for any reason.

TODO:
- Add `async.pcall`
- Provide a way for cancelled tasks to perform any cleanup and/or free resources.

Once that is done, is there any disagreement in using this for an initial MVP?

**Comment by lewis6991** [comment]
> Why does create return a thunk?

This is just a design decision to allow you to create async functions like:

```lua
local my_async_func = async.create(function(a, b, c, callback)
   ...
end, 4)
```

When you create something, you don't just execute it. Similar to `coroutine.create`, `uv.new_timer`, etc.

This is how I've used this lib in all the various plugins I've created and how I think the lib should be typically used.

> It should just run the async function in the coroutine

This is a subjective opinion without any reasoning. We can just as well create something like `async.run` if there's a real need. Not something I see myself using much.

> Why does join call the thunk with a callback? Are you supposed to give it a thunk the is returned from create?

You are supposed to pass `join` a list of async functions above. The callback is needed so the user can wait on the result.

```lua
local main = async.create(function()
   local results = async.join(math.huge, nil, {my_async_func1, my_async_func2})
   -- do something with results
end)
```

>What is this doing? Edit: it is for max number of concurrent tasks.
Why would that be useful?

Because some applications might not want to put too much strain on the event loop. I specifically wrote `join` to use in packer-v2, which has had many reports of issues when too many jobs are run in parallel.

> Why are we checking if the coroutine is running? We should just do what the function says it is doing, instead of changing depending on the context which is very confusing. We should only check if the coroutine is running to provide error messages to the user

It allows you to freely use functions wrapped in `async.create` in async contexts without creating more coroutines. Without it, it increases boilerplate.

**Comment by oberblastmeister** [comment]
> You are supposed to pass join a list of async functions above. The callback is needed so the user can wait on the result.

So you need to pass functions to join that are returned by `create`?
like this `async.join({async.create(f1, 1), async.create(f2, 2), async.create(f3, 1)}, math.huge, nil)`? That's pretty verbose IMO. I think `join` should create the coroutines, instead of asking the user to do so. Otherwise, you have to `async.create` every function that you want to `join`, which is more verbose, especially when you need to specify argc for each one. That is how it works in `plenary.nvim`.

> This is a subjective opinion without any reasoning. We can just as well create something like async.run if there's a real need. Not something I see myself using much.

Yes it is subjective, but I did this because most lua functions that run a closure in a context don't create a thunk. For example, `pcall`, `xpcall`, etc. If you want to create the thunk, you can do it yourself, `create` just does extra stuff that is not always needed.

> It allows you to freely use functions wrapped in async.create in async contexts without creating more coroutines. Without it, it increases boilerplate.

If you want to run functions in async contexts, just run the inner closure! I don't see any boilerplate. However, having functions expect different arguments depending on the context is very confusing. I also don't think leaf async functions should fallback to accepting a callback, they should just fail when not called in the proper context.

I also think we should try to remove the argc stuff. I think we can remove it as long as the callback is the first argument?

**Comment by rcarriga** [comment]
> This is a subjective opinion without any reasoning. We can just as well create something like async.run if there's a real need. Not something I see myself using much.

As you said this is completely subjective but I would have the opposite opinion. I've never wanted to use something like `create`. Of course one could be written using the other so it's not a big deal but might be worth considering which is best provided out of the box (if the choice is only one).

> Add async.pcall (as discussed aboved)

As one of the goals is to keep the core minimal, are you planning on keeping the `protected` argument for `wrap`? Seems like it's pretty niche, and could be implemented with `pcall` if desired.

> Why are we checking if the coroutine is running? We should just do what the function says it is doing, instead of changing depending on the context which is very confusing.

I agree with this. If I'm creating a new handle (by your naming) I'd expect it to create a new context (i.e. new coroutine). I don't think the core should try to be "smart". If an error occurs, the original async context also dies. This also means that if you're in an async function, you must block for all async calls made, it's not possible to start a task and not wait for it (and so actually `run` couldn't be written with `create`).

> It allows you to freely use functions wrapped in async.create in async contexts without creating more coroutines. Without it, it increases boilerplate.

I'd say this is a point towards not using `async.create`. Not defining functions wrapped with `create` leaves the user to decide how it's run.

**Comment by rcarriga** [comment]
> As one of the goals is to keep the core minimal, are you planning on keeping the protected argument for wrap? Seems like it's pretty niche, and could be implemented with pcall if desired.

Actually alternative idea is to add some extra logic to `wrap` to catch errors in wrapped functions and re-raise them in the coroutine that called them. That way no custom `pcall` is required, the built-in works as expected which I think would be a really nice addition

Implementation: https://github.com/rcarriga/nvim-dap-ui/blob/45ca04234a585708763b790986f588fcfc55fe52/lua/dapui/async/tasks.lua#L133

Some tests to demonstrate: https://github.com/rcarriga/nvim-dap-ui/blob/45ca04234a585708763b790986f588fcfc55fe52/tests/unit/async/tasks_spec.lua#L97-L113

**Comment by lewis6991** [comment]
> So you need to pass functions to join that are returned by create?

That was the intention, but actually just passing a callback style function will work.

> like this async.join({async.create(f1, 1), async.create(f2, 2), async.create(f3, 1)}, math.huge, nil)? That's pretty verbose IMO. 

The expectation is to do:

```lua

local f1 = async.create(function(...) end), 1)
local f2 = async.create(function(...) end), 2)
local f3 = async.create(function(...) end), 1)

async.join({f1, f2, f3)
```

Note: `join()` only works because `async.create` returns a callback-style function.

One of the main benefits of `async.create` is that it just returns a callback-style function, which means it can be used internally and exposed to the user without any API-dependency on the assync lib as it just becomes an implementation detail.

> If you want to run functions in async contexts, just run the inner closure! I don't see any boilerplate.

That requires you to create 2 versions of every function you want to run in both contexts. That was a problem I had in gitsigns, which was made worse by having a debounced version added to the mix. It simplified things a lot to just hide the async stuff, as it didn't matter.

>  However, having functions expect different arguments depending on the context is very confusing. I also don't think leaf async functions should fallback to accepting a callback, they should just fail when not called in the proper context.

To me the whole point of the async lib is that the async aspect is almost completely abstracted away. Wrapping a function in `async.create` is basically like adding `-- @async`. I'm simply saying this function **will** run in an async context, whether one is provided or not.

To me, it's similar to luv's API where every async function as a sync version where the callback is passed as `nil`.

For functions that don't return anything, they only difference is that one is blocking; they both execute the function as normal, no pesky unnecessary runtime errors. When there are return values, non-async returns its data via a callback, whilst the other returns normally. I don't see this as a significant enough difference to consider it confusing. I'd rather minimize the potential for runtime errors.

We could make it an error to pass a callback in the async-context? That would cover a good fraction of erroneous use cases.

> I also think we should try to remove the argc stuff. I think we can remove it as long as the callback is the first argument?

That would basically force every use of `wrap` to pass a closure, so I'm pretty against that. I've at least made `argc` optional (default to 0). We could extend that further and redefine `argc` as `callback_pos` so we can support arguments before **and** after the callback. Then it would be fully optional. Not sure if this would complicate things too much.

> As one of the goals is to keep the core minimal, are you planning on keeping the protected argument for wrap? Seems like it's pretty niche, and could be implemented with pcall if desired.

This area is a bit unfinished, but ideally we should aim to remove the argument and pass it some other way.

> I agree with this. If I'm creating a new handle (by your naming) I'd expect it to create a new context (i.e. new coroutine). I don't think the core should try to be "smart". If an error occurs, the original async context also dies. This also means that if you're in an async function, you must block for all async calls made, it's not possible to start a task and not wait for it (and so actually run couldn't be written with create).

> I'd say this is a point towards not using async.create. Not defining functions wrapped with create leaves the user to decide how it's run.

I'm trying to position the API as a little higher level to what It seems you're expecting, which is fine, maybe we can have both?

If we don't have `async.create` then we shouldn't have `async.wrap` either. If we follow this then we end up with:

```lua
function M.run(func, callback, ...)
  return execute(func, callback, ...)
end

-- Note: We could detect the type of arg 1 here. If number, then treat as argc, 
-- else treat as func and set argc = 0. The Lua stdlib does this kind of thing in a few places.
function M.wait(func, argc, ...)
  return coroutine.yield(argc, func, ...)
end
```

which is simpler than what I've got currently (though also a lot more limited).

However, we can then build `create` and `wrap` on top of this.

So best of both worlds?

> Actually alternative idea is to add some extra logic to wrap to catch errors in wrapped functions and re-raise them in the coroutine that called them. That way no custom pcall is required, the built-in works as expected which I think would be a really nice addition

This sounds like a good idea. I think it should work.

**Comment by lewis6991** [comment]
Updated https://github.com/lewis6991/async.nvim with all feedback:

- `protected` is gone. Used @rcarriga's implementation for re-raising errors in the coroutine.
- added `async.run` and `async.wait` to act as low level functions that execute closures.
  - additionally `async.wait` accepts an optional `argc` argument in the first position. Other Lua stdlib function use this kind of overloading (e.g. `debug.traceback`, `table.insert`) so I think it makes sense here.
- added a `strict` argument to `async.create`, `async.void` and `async.wrap` so users can opt-in to hard erroring when thunks execute in the wrong context. (I hope this is a good enough compromise).

**Comment by rcarriga** [comment]
> I'm trying to position the API as a little higher level to what It seems you're expecting, which is fine, maybe we can have both?

> - protected is gone. Used @rcarriga's implementation for re-raising errors in the coroutine.
> - added async.run and async.wait to act as low level functions that execute closures.
        additionally async.wait accepts an optional argc argument in the first position. Other Lua stdlib function use this kind of overloading (e.g. debug.traceback, table.insert) so I think it makes sense here.
> - added a strict argument to async.create, async.void and async.wrap so users can opt-in to hard erroring when thunks execute in the wrong context. (I hope this is a good enough compromise).

Awesome! Having the lower level access makes it much more flexible, so I think being more opinionated in functions like `create` or `join` is totally fine.

> Note: join() only works because async.create returns a callback-style function.

Though I think that should be well documented, it's a pretty large caveat.

**Comment by oberblastmeister** [comment]
> That would basically force every use of wrap to pass a closure, so I'm pretty against that.

To be clear, I am talking about the continuation that is given to `func`. If we make `wrap` (`wait`) defined as
```lua
function M.wrap(func, ...)
  return coroutine.yield(func, ...)
end
```
If `func` always takes the continuation as the first argument, then we do not need argc. So we should just define `wrap = coroutine.yield`. Then, if you want to pass in a `func` that expects the continuation to be the last argument, you can just do
```lua
wrap(function(k, arg1, arg2, arg3)
  f(arg1, arg2, arg3, k)
end)
```
You could also define a new wrap that takes an `argc`, etc. 

> I'm trying to position the API as a little higher level to what It seems you're expecting, which is fine, maybe we can have both?

This is neovim core, and we should focus on the fundamentals first.

**Comment by rcarriga** [comment]
> If func always takes the continuation as the first argument, then we do not need argc. So we should just define wait = coroutine.yield. Then, if you want to pass in a func that expects the continuation to be the last argument, you can just do

I don't really see the advantage of doing this, you're trading adding a single number parameter
```lua
wrap(vim.loop.fs_open, 4)
```
for writing a whole wrapper function
```lua
wrap(function (callback, path, flags, mode)
  vim.loop.fs_open(path, flags, mode, callback)
end)
```
Why is that better? Just simplifying the implementation a little doesn't seem worth it if almost every use has to write boilerplate (which is likely because callbacks are often the last argument)

**Comment by oberblastmeister** [comment]
It is better because it is the simplest and most general function.
That's why I said
> You could also define your own wrap that takes an argc.

Also, defining `wait` like that allows you to call `wait` on the fly with a closure without using `argc`, because it is basically just a control operation like `callcc` or something. You also avoid the overhead of doing an `unpack` and `pack`.

**Comment by lewis6991** [comment]
> It is better because it is the simplest and most general function.

If we want the simplest and most general then there's a stronger argument to just go with https://github.com/neovim/neovim/issues/19624#issuecomment-1204867814, and therefore we don't need to add anything to core! But that defeats the entire purpose here and I think we want to aim at a higher level.

If the suggested usage is going to force everyone to add a bunch of boilerplate and closures, then what's the point in adding anything to core? It's a low-cost, low benefit proposition. The point here is to standardize around well established patterns and reduce boilerplate, and respectfully I don't think they align with your suggestions, at least with the code I've seen.

> call wait on the fly with a closure without using argc

Yeah, but then you need to provide a closure. This is objectively worse.

> You also avoid the overhead of doing an unpack and pack.

That kind of overheads is negligible. Micro optimizing isn't a goal here and should not influence the API design unless the performance concerns are real.

**Comment by oberblastmeister** [comment]
> If we want the simplest and most general then there's a stronger argument to just go with https://github.com/neovim/neovim/issues/19624#issuecomment-1204867814, and therefore we don't need to add anything to core! But that defeats the entire purpose here and I think we want to aim at a higher level.

> If the suggested usage is going to force everyone to add a bunch of boilerplate and closures, then what's the point in adding anything to core? It's a low-cost, low benefit proposition. The point here is to standardize around well established patterns and reduce boilerplate, and respectfully I don't think they align with your suggestions, at least with the code I've seen.

I'm not saying we shouldn't have a high-level api, I am just saying that we should focus on the simplest and most general functions so that we can define the other functions *in terms* of those simple functions. We can still have `wrap`, `create`, etc, but they can be defined in terms of the low-level functions. You can't do it the other way around. That's what I am suggesting.

> Yeah, but then you need to provide a closure. This is objectively worse.

What do you mean? You have to provide a closure anyway. How is it objectively worse?

**Comment by lewis6991** [comment]
> I'm not saying we shouldn't have a high-level api, I am just saying that we should focus on the simplest and most general functions so that we can define the other functions in terms of those simple functions.

If you look at https://github.com/lewis6991/async.nvim you'll see that's exactly what I've done.

We define `wait` which accepts `argc` as an optional first parameter. If not provided it is assumed to be 1, meaning the continuation is the first argument. I think this will allow you to use closures as you suggested.

We also define `create` which is implemented with `run`. Again `argc` is optional.

> What do you mean? You have to provide a closure anyway. How is it objectively worse?

With `argc`:
```lua
local stat = async.wait(2, vim.loop.fs_stat, path)
```

Without `argc`:
```lua
local stat = async.wait(function(callback, path)
    vim.fs_stat(path, callback)
end)
```

**Comment by oberblastmeister** [comment]
Ok, the optional `argc` argument looks good. The current code doesn't actually set `argc` to 1 if it is not provided though. 

Everything basically looks good to me. I only have two small comments left.
1. I still think `join` is too verbose. Why can't it just create the coroutines itself without the user having to do that?
2. I think we should do the optional `argc` for `create` also.

**Comment by jtws** [comment]
Out of curiosity why was [promise-async](https://github.com/kevinhwang91/promise-async) not considered here? It implements async/await and full promises implementation with `promsie.all`, `promise.allSettled()`, `promise.race`, etc. and passes standard promise tests. I've been using it and prefer it to plenary's async barebones futures way.

In my opinion, the nvim solution shouldn't aim for simplicity but use established patterns and provide all the tools for async including synchronization, etc.

Here's  [Semaphore and mutex with promise-async](https://github.com/kevinhwang91/nvim-fundo/tree/main/lua/fundo/lib) CondVar wouldn't be much work to implement. The author has also made [async vim.loop wrapper using promises](https://github.com/kevinhwang91/nvim-fundo/blob/main/lua/fundo/fs/uvwrapper.lua) which I use.

You don't need Neovim-specific knowledge to use `promise-async` since it's modelled after JS so the web is full of guides and patterns for it.

Personally, I'll stick with it over any core implementation unless core features surpass those of `promise-async`

**Comment by miversen33** [comment]
@jtws 
This seems to be _roughly_ the answer to that question
https://github.com/neovim/neovim/issues/19624#issuecomment-1336298582

> Comforting javascript users is not a goal, and should not be factored into any of the designs here. As a typescript/javascript user myself, I'm very skeptical that async/await is a good interface. We should focus on what feels right for a Lua + Nvim.

**Comment by justinmk** [comment]
> In my opinion, the nvim solution shouldn't aim for simplicity but use established patterns and provide all the tools for async including synchronization, etc.

False dichotomy. We can have both, without blindly mimicking javascript.

> Python async uses futures and tasks and asyncio/asyncfile, JS promises, so why should Neovim core not aim for the same level of features and usability?

The "structured concurrency" features of python's tasks is in fact part of the goals described in the main issue description.

**Comment by svermeulen** [comment]
Some work was started on structured concurrency in lua (similar to Python's trio) [here](https://gist.github.com/belm0/4c6d11f47ccd31a231cde04616d6bb22) in case that approach might also be considered here

**Comment by svermeulen** [comment]
I was inspired enough by the article I linked to above that I created a plugin to provide structured concurrency in neovim lua:  https://github.com/svermeulen/nvim-lusc

**Comment by hrsh7th** [comment]
Should we support `progress` related functions in this area?

**Comment by s1n7ax** [comment]
Hello. I see there are bunch of cool implementations here. I'm just wondering if anyone has faced any issues using co-routines with asynchronous APIs, that is not a problem with using callbacks or promises approach?

Other than error handling? Nested co-routines are weird since it's resumes the last yielded co-routines. Any issues related to that?

**Comment by s1n7ax** [comment]
I have added Promise like error handling for https://github.com/ms-jpq/lua-async-await to handle LSP response type where the first value is the error. So no more assert checks on the response is needed. It's done in the `wait()` function too. So, if it's different, one can simply create a new wait function.

https://github.com/nvim-java/lua-async-await

```lua
local function lsp_request(callback)
  local timer = vim.loop.new_timer()

  assert(timer)

  timer:start(2000, 0, function()
    -- First parameter is the error
    callback('something went wrong', nil)
  end)
end

local M = require('sync')

local main = M.sync(function()
  local response = M.wait_handle_error(M.wrap(lsp_request)())
end)
  .catch(function(err)
    print('error occurred ', err)
  end)
  .run()
```

Result:

```
error occured  test6.lua:105: something went wrong
```

**Comment by vurentjie** [comment]
This is what I am currently using. 
https://gist.github.com/vurentjie/566a7158038ea6e044a4321c63cacde0 

It is very similar to some of the samples posted above in this thread, possibly because I also derived it from [/ms-jpq/lua-async-await](https://github.com/ms-jpq/lua-async-await) by breaking down all the calls and simplifying.

Thought I would just share it here as another example. It's very lean and probably does not cover everything. Currently I handle errors outside of this, either via `[p|xp]call()` or `error()` depending on the use case.

**Comment by CRAG666** [comment]
How to convert this vim.fn.jobstart "tectonic -X watch -x 'build --keep-intermediates --keep-logs'" to new vim.system?

**Comment by gregorias** [comment]
Hi,

I would like to present my design for structured concurrency in Neovim: [Coop.nvim](https://github.com/gregorias/coop.nvim). I believe it addresses all requirements you have listed out for such a framework + it’s as straightforward as such a framework could be ([How it works doc](https://github.com/gregorias/coop.nvim/blob/main/How%20it%20works.md)). I believe it provides an elegant approach to the subject.

Here’s how Coop relates to the listed requirements:

> Representing a task.
> Orchestrating "pipelines" (quasi monads?) of work ("tasks") and handling errors.

[See examples](https://github.com/gregorias/coop.nvim/blob/0e2082500707f2a143ff924ad36577c348148517/lua/coop/examples.lua#L72) for examples of handling errors, running tasks in parallel, and complex awaiting conditions.

> Maximally leveraging Lua coroutines + libuv. Only add concepts ("task", "promise") if absolutely needed.

Yes,

- the Task interface is an extension to Lua coroutines and implements parallel `create, resume, yield, status` functions with equivalent semantics.
- Libuv functions are easily ported through generic wrappers: https://github.com/gregorias/coop.nvim/blob/main/lua/coop/uv.lua.
- Only two concepts are added: a task and a future.

> Coroutines (or tasks that wrap coroutines) can be nested. ([ref](https://gregorias.github.io/posts/using-coroutines-in-neovim-lua/#addendum-whats-wrong-with-plenary-async))

True. In Coop you can freely nest task functions (a task function is a function that may call `task.yield`).

> Util to create an awaitable task from "normal" functions (cf. "promisify"?).
>  Example: vim.system() returns its own ad-hoc "task" that can be "awaited" via :wait().
> Can e.g. vim.system() be "promisified" without its knowledge?

Coop indeed “promisifies” `uv.spawn`. Coop’s implementation returns an awaitable future. [In tests you have an example of how it turns the callback interface from Neovim’s documentation into a synchronous one](https://github.com/gregorias/coop.nvim/blob/ff49c6d5a741e9e7b22366f5bedf43f7fcfd01c5/tests/coop/uv_spec.lua#L48-L84):

```lua
local stdin = vim.uv.new_pipe()
local stdout = vim.uv.new_pipe()
local stderr = vim.uv.new_pipe()

local handle, pid, cat_future = uv.spawn("cat", {
	stdio = { stdin, stdout, stderr },
})

local read_future = coop.Future.new()
vim.uv.read_start(stdout, function(err, data)
	assert(not err, err)
	if data ~= nil and not read_future.done then
		read_future:complete(data)
	end
end)
vim.uv.write(stdin, "Hello World")

-- Here we wrap the APIs into a task to use the synchronous looking interface.
local task = coop.spawn(function()
	-- Wait for stdout to produce the data.
	local read_data = read_future:await()
	assert.are.same("Hello World", read_data)
	local err_stdin_shutdown = uv.shutdown(stdin)
	assert.is.Nil(err_stdin_shutdown)
	-- Wait for the cat process to finish.
	return cat_future:await()
end)

-- Wait for 200ms for the task finish.
local exit_code, exit_sign = task:await(200, 2)
```

> Document (or generalize) ["coroutine to callback"](https://gregorias.github.io/posts/using-coroutines-in-neovim-lua/#coroutine-to-callback-conversion).

Coop provides a generic converter callback to coroutine converter: `cb_to_tf` ([docs](https://github.com/gregorias/coop.nvim/blob/main/How%20it%20works.md#using-callbacks-to-build-coroutines).)

The generic conversion from a coroutine (a task) into a callback is done through [an await overload that accepts a callback](https://github.com/gregorias/coop.nvim/tree/main#:~:text=task.await(function(success%2C%20result)%20end)).

> await_all, await_any (pseudo-names). See JS [Promise.all()](https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Global_Objects/Promise/all).
> Results (and failures) can be aggregated. (Can't do this with jobwait()!)

All possible. For example, Coop provides an implementation of `as_completed`, an iterator that returns task results as they finish: [example](https://github.com/gregorias/coop.nvim/tree/main#:~:text=sort_with_time%20shows%20that%20Coop%20achieves%20true%20parallelism.%20It%20launches%20parallel%20timers%20with%20coop.spawn%20and%20uses%20a%20coop.control.as_completed%20to%20conveniently%20capture%20results%20as%20each%20timer%20completes.).

Tentative implementations of `await_all` and `await_any` are in [control.lua](https://github.com/gregorias/coop.nvim/blob/ff49c6d5a741e9e7b22366f5bedf43f7fcfd01c5/lua/coop/control.lua#L1-L68).

> Tasks can be canceled.

[True.](https://github.com/gregorias/coop.nvim/blob/main/How%20it%20works.md#cancellation)

> Failures/errors can be handled

[True.](https://github.com/gregorias/coop.nvim/blob/main/How%20it%20works.md#error-handling)

> (possibly canceling the rest of the task tree).

Coop doesn’t natively provide the task tree abstraction to keep with the Lua’s spirit of coroutines being standalone threads. Task trees can be implemented on top of Coop, because the programmer can intercept the `"cancelled"` error and `:cancel` subtasks.

**Comment by justinmk** [comment]
@gregorias Thanks for sharing, that looks very promising.

**Comment by brianhuster** [comment]
I found out that Python's [toolz](https://toolz.readthedocs.io/en/latest/api.html#toolz.functoolz.pipe) library has a `pipe()` function that just take a value and then function(s) as argument. Is it something wanted in Nvim stdlib?

**Comment by dmitmel** [comment]
On the matter of cancellation, what do you think of the `CancellationToken` abstraction from VSCode? The idea is as follows: the asynchronous functions get an object that lets you subscribe to a cancellation event and register a callback to cancel whatever task you are implementing. The subscription to this event gives you back a callback to unsubscribe, say, after the request has been fulfilled, to not keep the referenced objects around in memory. When a cancellation is detected, the function it was called from should return some null value, which will be detected and propagated by the caller function and so on. In pseudocode, this would look something like this:

```lua
local function lsp_hover(client: lsp.Client, token: async.CancellationToken)
  local req = client.request('textDocument/hover') -- Let's imagine that client.request is a fafcf
  local unsubscribe = token.subscribe(function() req:cancel() end)
  local err, res = req:await()
  unsubscribe()
  if token.cancelled then
    return nil
  end
  -- All good, do something useful
end
```

To cancel the request one would store the token somewhere and call `cancel()` on it, or something along those lines, i.e.:

```lua
local token = async.CancellationToken.new()
autocmd('CursorMoved', function() token:cancel() end)
lsp_hover(client, token)
```

I chose the LSP hover method in this example purely because incidentally I was working on something similar, this really can be any asynchronous API. As far as I can tell from my experience with the codebase of VSCode and coc.nvim (which only involves reading code, not writing anything, but still) this kind of cancellation API jives really well with javascript's async/await style concurrency, which I think can be more-or-less mapped 1-to-1 to fafcfs. Its main benefit is in the fact that once an API supports a cancellation token, its application is a matter of calling a cancellable function and checking if its return value is `nil` or `token.cancelled` is true, e.g.:

```lua
local function lsp_hover(client: lsp.Client, token: async.CancellationToken)
  -- The subscription and unsubscription are encapsulated in `client.request`
  local err, res = client.request(token, 'textDocument/hover'):await()
  if not err and not res then -- The request must've been cancelled, propagate this info
    return nil
  end
  -- All good, do something useful
end
```

Mixing this pattern with error handling is TBD.

P. S. As for fafcf-style async in Lua in general, here's a short helper I came up with for my personal plugins:

```lua
--- Helper for turning callback-based APIs into asynchronous functions. Based on
--- <https://github.com/gregorias/coerce.nvim/blob/4ea7e31b95209105899ee6360c2a3a30e09d361d/lua/coerce/coroutine.lua>
--- and <https://gregorias.github.io/posts/using-coroutines-in-neovim-lua/>.
---@param fn fun(callback: fun(...))
---@return ...
function M.await(fn)
  local co = assert(coroutine.running(), 'needs to be called within a coroutine')
  local results = nil

  local function async_callback(...)
    results = { n = select('#', ...), ... }
    if coroutine.status(co) == 'suspended' then
      local ok, err = coroutine.resume(co)
      if not ok then error(debug.traceback(co, err)) end
    end
  end

  fn(async_callback)

  if not results then coroutine.yield() end
  assert(results, 'async callback did not get called')
  return unpack(results, 1, results.n)
end
```

**Comment by lewis6991** [comment]
The mechanism I'll be putting forward will adopt the method python asyncio handles this. There will be task objects with a `cancel()` method.

**Comment by justinmk** [comment]
> There will be task objects with a `cancel()` method.

Yeah, a single "task" abstraction sounds nice, rather than a disconnected "cancel token". Ideally it should be possible to create an empty task object (without actually invoking a job or anything), and pass that around to serve as a "cancel token" if needed.

**Comment by lewis6991** [comment]
That sounds like a future (which a task builds on)? I've implemented that too along with other primitives: event, queue, semaphore, etc.

**Comment by lewis6991** [comment]
I've raised #34473 to initial feedback gathering.

**Comment by emiasims** [comment]
I had an implementation based on `coop.nvim` by @gregorias (i.e. largely verbatim), but also added an async pipeline similar to `Iter`, complete with tests, at https://github.com/emiasims/neovim/tree/task .

`vim.pipe`, looks like I can translate it very easily to the implementation in #34473, and my `vim.co` is less complete than the PR added. If it makes sense, I will adapt `vim.pipe` to lewis' implementation and submit a draft PR.

**Comment by justinmk** [comment]
@emiasims `vim.pipe` is a very interesting idea, please open a new issue to discuss it (and link to this one). At a high level I would ask if `vim.iter` and/or `vim.async`/`vim.task` could serve that purpose, why do we need a separate `vim.pipe` interface? (But let's discuss that in the new issue) The general idea of `vim.iter` and `vim.async` (or `vim.task`) is to be able to compose tasks.

**Comment by aktau** [comment]
Just saw cf0f90fe14f5d806be91d5de89d04c6821f151b7 (`vim._async` by @lewis6991) being merged into master. I assume also to test drive APIs.

**Comment by justinmk** [comment]
Just read https://lucumr.pocoo.org/2025/7/26/virtual-threads/ (suggested by @ii14), IIUC the key idea is:

- Thread group = a "root thread" with child threads.
  - Awaiting the thread group awaits all child threads.
  - Cancelling the group cancels all child threads.
  - If any child thread fails, future spawns are blocked and existing threads are requested to cancel.

From discussion, it sounds like this is how https://github.com/neovim/neovim/pull/34473 behaves too.

**Comment by lewis6991** [comment]
Yes, the model of cancellation and groups is the main thing I'm still working on (and why the pr is still draft).

My main thinking is to implement several behaviours of trio / anyio (python) directly as part of tasks which is the same as the points you just described.

- tasks created within another task have a child/parent relationship
- parent tasks can't finish until all children have
- cancellations propagate downwards to all children
- children can detect cancellations so they can perform cleanup
- cancellations are level triggered (as opposed to edge triggered like in pythons asyncio)

The means:
- children are never orphaned
  - https://vorpus.org/blog/notes-on-structured-concurrency-or-go-statement-considered-harmful/ convinced me that thisnis the right approach 
- children can't easily swallow cancellations (edge triggered vs level triggered)

> If any child thread fails, future spawns are blocked and existing threads are requested to cancel.

I'm not so sure what this one means exactly. 

You don't know if a child thread has failed until you wait for it to complete. In async/await, if the task fails then the error will be thrown on the `await` call, which can be guarded with a try-catch (or `pcall` in Lua). And additionally in vim.async, an error occurring in the parent thread will cause all child tasks to be cancelled.

**Comment by hrsh7th** [comment]
In my opinion, we should use the Linux Process model as a reference to implement and explain parent-child relationships and cancellation handling.

I think this would be a natural metaphor to understand, without using new terms like nursery.

**Comment by lewis6991** [comment]
Yes, this is effectively what has been implemented, and influenced by your proposal.

Basically all async contexts act as a nursery rather than it being a separate thing. The goal is that children are never orphaned unless explicitly detached.

**Comment by aktau** [comment]
I'm curious about this example: https://github.com/neovim/neovim/blob/36cc62b27633f170d895d66b230eb5dd66b9ca1a/runtime/lua/vim/async.lua#L162-L180.

I assume the "detaching" happens because that is documented to "start the work immediately" (by virtue of top-level tasks starting immediately). But semantically these tasks do belong to the one that created them (in the structured concurrency sense). Wouldn't `vim.async.iter()` also start running all of them? If not, is there any way to do this without breaking parent/child link?
